compiler/src/iree/compiler/Codegen/Common/GPU/GPUCombineValueBarriers.cpp

```diff
@@ -0,0 +1,268 @@
+  //  insertion point,
+  //  - Any producers will still remain before the operation, as we are only
+  //  moving the operation after.
+  for (Operation *sliceOp : llvm::reverse(toMove)) {
+    rewriter.moveOpAfter(sliceOp, insertionPoint);
+  }
+
+  return success();
+}
+
+/// Combine all value barriers into a single value barrier.
+static LogicalResult
+combineValueBarrierOps(RewriterBase &rewriter, Location loc,
+                       ArrayRef<IREE::GPU::ValueBarrierOp> valueBarriers) {
+  if (valueBarriers.size() <= 1) {
+    return success();
+  }
+  SmallVector<Value> barrierOperands;
+  for (auto barrierOp : valueBarriers) {
+    barrierOperands.append(barrierOp.getInputs().begin(),
+                           barrierOp.getInputs().end());
+  }
+  auto combinedBarrierOp =
+      rewriter.create<IREE::GPU::ValueBarrierOp>(loc, barrierOperands);
+
+  // Replace all uses of the previous barrier with new barrier.
+  int resultNumber = 0;
+  for (auto barrierOp : valueBarriers) {
+    int numResults = barrierOp.getNumResults();
+    rewriter.replaceOp(barrierOp, combinedBarrierOp->getResults().slice(
+                                      resultNumber, numResults));
+    resultNumber += numResults;
+  }
+  return success();
+}
+
+/// Given two barriers, barrierA and barrierB, combine them into a single
+/// barrier.
+static FailureOr<IREE::GPU::ValueBarrierOp>
+combineValueBarrierPair(RewriterBase &rewriter,
+                        IREE::GPU::ValueBarrierOp barrierA,
+                        IREE::GPU::ValueBarrierOp barrierB) {
+  // Both barriers need to have either tensor semantics or vector semantics.
+  if (barrierA.hasTensorSemantics() && !barrierB.hasTensorSemantics()) {
+    return failure();
+  }
+  if (!barrierA.hasTensorSemantics() && barrierB.hasTensorSemantics()) {
+    return failure();
+  }
+
+  // We assume barrierA is always before barrierB.
+  if (barrierB->isBeforeInBlock(barrierA)) {
+    std::swap(barrierA, barrierB);
+  }
+
+  // barrierA and barrierB are in the same block.
+  Block *block = barrierA->getBlock();
+
+  auto sliceFilter = [&block](Operation *candidate) -> bool {
+    if (candidate->getBlock() != block) {
+      return false;
+    }
+    if (candidate == block->getTerminator()) {
+      // Do not move the terminator.
+      return false;
+    }
+    return true;
+  };
+
+  // Find the combined backward slice of barrierA and barrierB and try
+  // to move it before barrierA (before both the barriers).
+  BackwardSliceOptions bOptions;
+  bOptions.filter = sliceFilter;
+  SetVector<Operation *> backwardSliceA;
+  SetVector<Operation *> backwardSliceB;
+  getBackwardSlice(barrierA, &backwardSliceA, bOptions);
+  getBackwardSlice(barrierB, &backwardSliceB, bOptions);
+  backwardSliceA.insert(backwardSliceB.begin(), backwardSliceB.end());
+  // Move the backward slice before barrierA.
+  if (failed(moveBackwardSliceBeforeOp(rewriter, backwardSliceA, barrierA))) {
+    return failure();
+  }
```

Comment:
failures from moveBackwardSliceBeforeOp/moveForwardSliceAfterOp get propagated up and returns failure (see comment on each of these functions). This is the actual check which leads to failure when both barriers have a common ancestor.